Upgrade to mypy 0.720

pradyunsg opened the pull request:
<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

cjerdonek commented on src/pip/_internal/utils/appdirs.py:
```diff
@@ -239,7 +239,9 @@ def _get_win_folder_with_ctypes(csidl_name):
-    return buf.value
+    # On Python 2, ctypes.create_unicode_buffer().value returns "unicode",
+    # which isn't the same as str in the annotation above.
+    return buf.value  # type: ignore
```
I would convert this to `str` before returning. There are already similar functions or lines of code in the code base, but I would first try converting using `sys.getfilesystemencoding()` and then fall back to other encodings if that doesn't work. I think it's important for this type annotation to be correct, otherwise it can lead to surprising `UnicodeError`s, etc. in the calling code since people looking at the types will be expecting `str` since that's what they'll see (and the type checker won't catch it either).

pradyunsg replied:
I share the concern here -- however given my lack of experience w/ unicode issues and that this is Python 2 only, I personally *really* don't want to spend time working on it. 🙃

Plus it's in appdirs' internal copy so making those fixes should help too.

cjerdonek replied:
Okay, then I would at least add a code comment near the function’s type annotation warning that it doesn’t return str under Python 2. Even if we don’t want to work on it, it’s still good to take steps to avoid introducing bugs when we’re touching the code.

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -65,9 +65,9 @@
     # typing's cast() isn't supported in code comments, so we need to
     # define a dummy, no-op version.
-    def cast(typ, val):
+    def cast(typ, val):  # type: ignore
```
I would add a code comment saying it's okay to ignore the type info here since if we're in the except block, we know we're not in the process of checking types (which isn't true for all such `ImportError` except blocks).

pradyunsg replied:
TBH, I want to just change this up to get rid of the "from typing import ..." that may execute at runtime and add an else clause to the if mypy check above.

I'd prefer to do that in a follow up though.

pradyunsg replied:
I'll add this change to this PR, since it's pretty trivial.

cjerdonek commented on src/pip/_internal/pep425tags.py:
```diff
@@ -353,6 +354,7 @@ def get_supported(
 
         # Current version, current API (built specifically for our Python):
         for abi in abis:
+            assert abi
```
Does it work to instead annotate `abi3s` above with `Set[str]` (or similar)? Then you can rely on the type checker without having to do asserts.

pradyunsg replied:
This file's gonna go away soon enough, when we switch to packaging.tags.

Thus, I'm not too concerned about this.

---

The issue here is that there's an abi attribute in the parameters of the function which is optional string and that name is being reused.

cjerdonek replied:
Okay, then I would add a code comment that the assert is needed to satisfy the type checker. Otherwise, it’s not clear to someone reading the code why it’s there.

pradyunsg replied:
Sounds fair to me. I'll add one.

pradyunsg replied:
> Does it work to instead annotate `abi3s` above with `Set[str]` (or similar)?

AH! This works. Please don't mind me. 🙈

pradyunsg commented:
Thanks for the feedback @cjerdonek!

I'm going to go ahead and take the liberty of merging this PR now, since I think I've addressed your comments here. If you have additional inputs, I'll be happy to file a new PR to follow up on those.

Please do let me know if you mind this. :)

cjerdonek commented:
Thanks, looks great!
